Improve predictability of writeUtf8/writeAscii performance

franz1981 opened the pull request:
Motivation:

writeUtf8 can suffer from inlining issues and/or megamorphic call-sites on the hot path due to ByteBuf hierarchy

Modifications:

Duplicate and specialize the code paths to reduce the need of polymorphic calls

Result:

Performance are more stable in user code

franz1981 commented:
Need to explain the context of this PR with https://github.com/eclipse-vertx/vert.x/pull/3462#issuecomment-647038310: when user code mix different `ByteBuf` subtypes or just has it's own stack depth, the `-XX:MaxInlineLevel` config (9 on jdk 8 and 15 on jdk 14) could prevent certain calls to be inlined causing unwanted bound checks, prevent vectorizations...hitting the stability of performance we see in the micro-benchmarks or just in an ideal world.

This duplication is aiming to makes the performance a bit more stable (sadly `PlatformDependent::putByte` has still some depth and `CharSequence:.charAt` too, but the benchmarks already shows some improvement.
I don't think I will left the benchmark changes to be there in the final PR because are too tied to how the JDK is configured.

franz1981 commented:
I've tweaked the benchs and impl to shows what's the effects of these changes...
this PR:
```
Benchmark                             (direct)   Mode  Cnt        Score        Error  Units
ByteBufUtilBenchmark.nestedWriteUtf8      true  thrpt   20  1135949.136 ±  15508.180  ops/s
ByteBufUtilBenchmark.nestedWriteUtf8     false  thrpt   20  1115468.247 ±  65425.276  ops/s
ByteBufUtilBenchmark.writeUtf8String      true  thrpt   20  8237112.811 ± 274131.929  ops/s
ByteBufUtilBenchmark.writeUtf8String     false  thrpt   20  4164310.459 ± 117297.047  ops/s
```
Before this PR with `-XX:MaxInlineLevel=9`:
```
Benchmark                             (direct)   Mode  Cnt        Score        Error  Units
ByteBufUtilBenchmark.nestedWriteUtf8      true  thrpt   20   738208.471 ±  13030.643  ops/s
ByteBufUtilBenchmark.nestedWriteUtf8     false  thrpt   20   815001.393 ±  48192.092  ops/s
ByteBufUtilBenchmark.writeUtf8String      true  thrpt   20  5859888.733 ± 143176.688  ops/s
ByteBufUtilBenchmark.writeUtf8String     false  thrpt   20  4484238.436 ±  64584.530  ops/s
```
Before this PR with `-XX:MaxInlineLevel=15`:
```
Benchmark                             (direct)   Mode  Cnt        Score       Error  Units
ByteBufUtilBenchmark.nestedWriteUtf8      true  thrpt   20  5499785.717 ± 59616.413  ops/s
ByteBufUtilBenchmark.nestedWriteUtf8     false  thrpt   20  4954847.853 ± 70535.926  ops/s
ByteBufUtilBenchmark.writeUtf8String      true  thrpt   20  5826841.982 ± 66342.052  ops/s
ByteBufUtilBenchmark.writeUtf8String     false  thrpt   20  4597271.515 ± 17988.082  ops/s
```
on jdk8 `nestedWriteUtf8` shows what's the possible effects of failing to inline some of the Netty `buffer` calls (or `PlatformDependent` too), while `String::charAt` isn't inlined (sadly, but could be fixed too on jdk 8 with some "hack").
`writeUtf8String` shows instead the effect of this PR while everything is correctly inlined instead, and performance are slightly improved too ;)
In this benchmark I haven't shows what's the effect of polymorphism (bi,tri, mega...) but I'm sure is needed, because just inlining shows the motivation behind the duplicated code.

franz1981 commented:
I've pushed a further commit that is manually inlining the slow path of `writeUtf8` ie `writeUtf8Surrogate` in order to make the methods > `-XX:FreqInlineSize` (ie 325 bytes)
eg `io.netty.buffer.ByteBufUtil::unsafeWriteUtf8 (358 bytes)   hot method too big` and 
`io.netty.buffer.ByteBufUtil::arrayWriteUtf8 (328 bytes)   hot method too big`
The results seems nice:
```
Benchmark                             (direct)   Mode  Cnt        Score        Error  Units
ByteBufUtilBenchmark.nestedWriteUtf8      true  thrpt   20  6524381.794 ± 102816.039  ops/s
ByteBufUtilBenchmark.nestedWriteUtf8     false  thrpt   20  4101033.546 ± 238837.837  ops/s
ByteBufUtilBenchmark.writeUtf8String      true  thrpt   20  7958679.895 ± 687717.569  ops/s
ByteBufUtilBenchmark.writeUtf8String     false  thrpt   20  4502402.790 ±  91698.297  ops/s
```
But I would like to see how the ASM looks like

franz1981 commented:
I'm not very used to achieve good numbers by playing so much with the JDK implementation details: I would like some good eyes to help me a bit on this (@nitsanw if he has the chance to look at this)

franz1981 commented:
In the last commit I've tried to use `PlatformDependent` on `byte[]` to save the array bound-checks to happen and indeed I see some improvement: it couldn't be used before becaure now the `xxxWriteUtf8` methods are not inlineable anymore...
```
Benchmark                             (direct)   Mode  Cnt        Score        Error  Units
ByteBufUtilBenchmark.nestedWriteUtf8      true  thrpt   20  7848114.042 ± 229818.691  ops/s
ByteBufUtilBenchmark.nestedWriteUtf8     false  thrpt   20  5907367.968 ± 321506.155  ops/s
ByteBufUtilBenchmark.writeUtf8String      true  thrpt   20  8264472.002 ±  49110.805  ops/s
ByteBufUtilBenchmark.writeUtf8String     false  thrpt   20  6166188.985 ± 336638.483  ops/s
```

franz1981 commented:
The thing that isn't 100% convincing me is the magnitude of error that doesn't seem proportional to the gain: I still need to investigate what's going on.

franz1981 commented:
I suppose @trustin could be interested into this as well :)

nitsanw commented:
I know the code is based off of some of the hacks I did on this blog post : http://psy-lob-saw.blogspot.com/2012/12/encode-utf-8-string-to-bytebuffer-faster.html

But it's been a while :-), reading through the patch though, I recall having an implementation somewhere that used the same code for array and memory case, there should be no need for the code duplication. Also, the blog post had a better data set to use for the benchmark IIRC.

njhill commented:
I've [commented in past](https://github.com/netty/netty/pull/9260#issuecomment-504594301) how sensitive `ByteBuf` access is in general to stack-depth due to transitive optimizations from inlining (eliminating accessibility checks). It means that in cases like this where there are many granular accesses, how deep the call stack happens to be makes an enormous difference to performance.

Having started to go down the rabbit hole of manually inlining parts of the `ByteBuf` hierarchy I came to the conclusion that it's not really worth trying to trim the stack depth by one or two calls for specific applications if that results in a lot of duplication and less-maintainable code, since it only helps in the borderline cases (the app in question might add an additional level of method indirection, at which point the benefit is lost).

Instead I feel we should just strongly advertise the fact that users can often get a big performance boost by setting `-XX:MaxInlineLevel` much higher than the default (at least for java < 14). I haven't encountered any cases where doing so is detrimental.

But otherwise I do really like the idea of speeding up reading/writing utf8 to `ByteBuf`s if/where possible given its prevalence on critical paths.

One specific question about the changes is why inlining `writeUtf8Surrogate` makes sense given how rarely that path would be hit?

franz1981 commented:
@njhill 
> One specific question about the changes is why inlining writeUtf8Surrogate makes sense given how rarely that path would be hit?

https://github.com/netty/netty/pull/10368#issuecomment-648217848

Tldr to save the big method from being inlined and "refill" the inlining budget: this would save from being dependent by the user code stack depth and would make any PlatformDependent call inlineable (because < 9th stack depth).
By consequence, CharSequence::chatAt became a bimorphic call for most usages (String or StringBuilder), getting inlined as well.

njhill commented:
Thanks @franz1981, I guess maybe there's something I misunderstood about the inlining budget. I thought `MaxInlineLevel` was based on absolute stack depth (at bytecode level), so whether a parent method is inlined or not wouldn't affect the effective stack depth of its callees for this purpose. If you are saying that the depth counter gets reset to zero from any non-inlined method in the stack I would be very interested and happy to be mistaken! There's probably a bunch of other places that could benefit from similar parent method "inline blocking".

Though I don't follow why this affects whether `charAt` is considered bimorphic or not, and also don't follow why it's necessary to split into the various similar implementations since the main thing that matters afaik is that the max inlining depth is sufficient to collapse all of the descendent `ByteBuf` accessibility checks into the same level.

franz1981 commented:
@njhill 
> If you are saying that the depth counter gets reset to zero from any non-inlined method in the stack I would be very interested and happy to be mistaken!

:) TBH I have tried it by writing some tests (similar to the one shown on this PR, with the nested calls), but that's why I needed the @nitsanw validation on this ;)

I'll try tomorrow to answer you back to the other questions you've rightly asked: I see that I have pushed into this PR much motivations that I haven't explained clearly :P

nitsanw commented:
Both @franz1981 and @njhill are pointing to concrete factors in inlining considerations. Roughly speaking we have:
1. inlining depth (can configure, but people rarely do, I support Nick's recommendation and have seen in applied widely to good result), too deep and it won't get inlined. This depth is computed from the perspective of the method being compiled, not the base of the stack.
2. method size (also configurable), too large and it won't get inlined
3. failure to de-virtualize, e.g. due to a mega-morphic callsite

IIUC Franz is trying to improve inlining by manually unrolling call-sites and by creating a method large enough to **prevent** inlining. If you prevent inlining of method X you are effectively resetting the inlining budget/scope and allowing the compiler a better chance of inlining it's callees into it.

Shaping compilation at this level without inlining control (e.g. via CompileCommand) is brittle, but can be effective.

I hope this helps in clarifying the challenge/scope/tools :-)

franz1981 commented:
Nitsan has perfectly summarised and understood the intention and motivation of this PR, thanks :)

Nonetheless he has confirmed the "inlining budged refill hack": the reason why I have performed tests comparing performance vs a tweaked MaxInliningLevel on 4.1 is that this PR should be at least on par (or better) then the original version regardless the hack to represent a step forward. 
Most of this come from the code duplication that's peeling off the callsites to make each writeUt8 a monomorphic version (related to ByteBuf) of the original code, decoupled from the ByteBuf hierarchy.

nitsanw commented:
https://github.com/nitsanw/jmh-samples/blob/master/src/main/java/psy/lob/saw/utf8/Utf8EncodingBenchmark.java

The benchmark I mentioned, which can still be improved

franz1981 commented:
I've adapted (ie copied) the nice benchmark data of @nitsanw into a new benchmark and these are the results:
Post PR:
```
Benchmark                                                (direct)   Mode  Cnt      Score     Error  Units
Utf8EncodingBenchmark.byteBufUtilWriteUtf8String             true  thrpt   20  13865.136 ± 262.175  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8String            false  thrpt   20  12147.668 ± 426.339  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8StringBuilder      true  thrpt   20  12688.814 ±  38.461  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8StringBuilder     false  thrpt   20  11801.796 ± 147.851  ops/s
Utf8EncodingBenchmark.nestedByteBufUtilWriteUtf8String       true  thrpt   20  12943.926 ± 133.867  ops/s
Utf8EncodingBenchmark.nestedByteBufUtilWriteUtf8String      false  thrpt   20  10628.178 ±  98.822  ops/s
```
Pre PR:
```
Benchmark                                                (direct)   Mode  Cnt      Score     Error  Units
Utf8EncodingBenchmark.byteBufUtilWriteUtf8String             true  thrpt   20  12712.676 ±  45.297  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8String            false  thrpt   20  13256.125 ± 148.626  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8StringBuilder      true  thrpt   20  12622.236 ±  29.238  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8StringBuilder     false  thrpt   20  13119.969 ± 252.603  ops/s
Utf8EncodingBenchmark.nestedByteBufUtilWriteUtf8String       true  thrpt   20   3000.798 ±  24.573  ops/s
Utf8EncodingBenchmark.nestedByteBufUtilWriteUtf8String      false  thrpt   20   2500.261 ± 343.188  ops/s
```
I see that array based `ByteBuf` are now getting a small perf hit, while direct ones are improved: I need to go down into the rabbit hole to understand what's going on.

`nestedByteBufUtilWriteUtf8String` probably won't survice the PR when/if merged because too tied to the JDK version, but it shows the reason behind this PR: we're nearly getting a 5X improvement (!).

franz1981 commented:
@njhill I think we have now some nice tool to play with :)
@njhill @normanmaurer @nitsanw
As a general rule of thumb, well optimized code always tends to play dirty with the JVM rules and we often try to pack cold/exceptional cases/branches away from the expected hottest path to make things inlineable. This PR is doing the same but for the opposite reason hence IMO it's fragile as much as the code it's going to replace. 

The other optimization to reduce the influence of the ByteBuf hierarchy depth is performed in the form of the code duplication, that Nitsan correctly said could be saved for byte[] and off-heap case, given that from the Unsafe perspective they are the same, but it would require me to modify PlatformDependent. This last optimization is what makes the code uglier, but is probably what's making the direct ByteBuf case faster too.

franz1981 commented:
Re 
> I see that array based ByteBuf are now getting a small perf hit

I see on the ASM that I'm getting some unexpected bound checks although using `PlatformDependent`, while if I use directly Unsafe and store in a local field the cached byte[] array offset I get back the original performance (and no bound checks):
```
Utf8EncodingBenchmark.byteBufUtilWriteUtf8String           false  thrpt   20  13163.882 ± 160.442  ops/s
```

franz1981 commented:
The latest commit shows the best performance (if compared to 4.1 too):
```
Benchmark                                               (direct)   Mode  Cnt      Score     Error  Units
Utf8EncodingBenchmark.byteBufUtilWriteUtf8String            true  thrpt   20  13632.727 ± 350.050  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8String           false  thrpt   20  13744.658 ±  86.133  ops/s
Utf8EncodingBenchmark.nestedByteBufUtilWriteUtf8String      true  thrpt   20  11753.634 ± 326.259  ops/s
Utf8EncodingBenchmark.nestedByteBufUtilWriteUtf8String     false  thrpt   20  12387.934 ± 171.924  ops/s
```

normanmaurer commented:
@chrisvest may be interesting to you as well

njhill commented:
Thanks @franz1981 @nitsanw for the very helpful additional clarifications and sorry for delayed reply, makes more sense to me now and I learnt something new. The megamorphic call-site concern is only w.r.t. `ByteBuf` rather than `CharSequence`, right?

I think the changes look great for the most part. Couple of other thoughts...
- Instead of subtlety inflating the size of each of those methods to block inlining, how about just doing so explicitly in the parent method (i.e. top-level `writeUtf8`)? Then the `writeUtf8Surrogate` method could be de-duplicated again (by adding a `ByteBuf` to arg to the unsafe variants)
- I wonder if the artifical inline-blocking could actually harm some use-cases, for example where the stack isn't very deep to start with and there are a large number of repeated calls to `writeUtf8` with small char sequences?

Just thinking aloud but maybe it's possible to make this trick a bit more adaptive, where the stack depth is detected the first time that it's called and stored in a (plain) static flag. This and/or programmatic inspection of `-XX:MaxInlineLevel` could then be used to decide whether or not to pass through an artificially inflated inline-blocking method. I guess that might not work in cases like this where the stack depth could be very different between different calls in the same app.

franz1981 commented:
@njhill 
>  The megamorphic call-site concern is only w.r.t. ByteBuf rather than CharSequence, right?

Yes and it's one of the motivations for this PR (the primary was the inlineability of `ByteBuf` & `PlatformDependent` calls; that's related). Often people think that `PlatformDependent` calls are nearly free, like using `Unsafe` but the reality is that `Unsafe` methods are intrinstics while `PlatformDependent` ones still play a role in the inlining dance: this PR is aiming to save having a whole method call while accessing a single byte.
A negative note on this P, instead,  is related to potential magamorphism of `CharSequence::charAt` that could became a concern now, given that each specialized version of `writeUtf8` isn't inlineable, `charAt` could see several implementors of `CharSequence` because the compiled method is shared across different callers AFAIK; but if there is a dominant type and the users will use it with < 3 types (String, StringBuilder?) I see JIT is still able to produce decent code ie hoisting the type check out of the loop and duplicating loop bodies to cover the bimorphic cases.
With >=3 I still need to check (and with JDK 11 is a real guess, given that `String` has got a wider hierarchy IIRC).

> - Instead of subtlety inflating the size of each of those methods to block inlining, how about just doing so explicitly in the parent method (i.e. top-level writeUtf8)? Then the writeUtf8Surrogate method could be de-duplicated again (by adding a ByteBuf to arg to the unsafe variants)

That's an option: I've tried to have 3 different method types to help with `CharSequence` polymorphism: each not-inlineable variant has it's own seen types on `charAt` level

 > I wonder if the artifical inline-blocking could actually harm some use-cases, for example where the stack isn't very deep to start with and there are a large number of repeated calls to writeUtf8 with small char sequences?

Would be nice to have a real case that heavily relies on `writeUtf8`: micro benchs sees that writing small char sequences can have a perf hit, but in the real world these ones still need to access some memory that could be cold, getting a cache miss on the first access: it should be dominant if compared with a full-fat method call, but only a real world case could help to understand it. Feel free to look at my benchs (borrowed from the nice ones of @nitsanw).

> Just thinking aloud but maybe it's possible to make this trick a bit more adaptive, where the stack depth is detected the first time that it's called and stored in a (plain) static flag. This and/or programmatic inspection of -XX:MaxInlineLevel could then be used to decide whether or not to pass through an artificially inflated inline-blocking method. I guess that might not work in cases like this where the stack depth could be very different between different calls in the same app.

Ah it would be rather easy if we could control the inlineability of `PlatformDependent` too :) 
But yes: i just wanted to not reduce dramatically the quality of the original code, using the same "trick" we use to get inlineable calls; I'm sure we could further improve it by detecting somehow the context by which the user is using this, but only if it worths (see previous points: I don't know yet in a real-case scenario with several small written strings how this PR would negatively impact the perf)

franz1981 commented:
@njhill 
In order to show my concern about the >2 case, I've modified the benchmark to shows the effect of bi/polymorphic cases of `CharSequence` args, results below:

this PR:
```
Benchmark                                              (direct)  (noUnsafe)   Mode  Cnt      Score     Error  Units
Utf8EncodingBenchmark.byteBufUtilWriteUtf8Bimorphic       false       false  thrpt   10   4738.091 ±  40.236  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8Bimorphic       false        true  thrpt   10   3784.695 ± 114.724  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8Bimorphic        true       false  thrpt   10   4758.198 ±  66.403  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8Bimorphic        true        true  thrpt   10   2242.832 ±  61.926  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8Megamorphic     false       false  thrpt   10   1271.008 ±  22.136  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8Megamorphic     false        true  thrpt   10   1168.745 ±   6.868  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8Megamorphic      true       false  thrpt   10   1250.535 ±  59.894  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8Megamorphic      true        true  thrpt   10    991.959 ±   4.620  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8String          false       false  thrpt   10  10486.611 ± 192.794  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8String          false        true  thrpt   10   7484.192 ± 160.313  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8String           true       false  thrpt   10  10719.207 ± 104.911  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8String           true        true  thrpt   10   4532.280 ± 280.026  ops/s
```
still awaiting 4.1 results

chrisvest commented:
Hi @franz1981,

If I understand correctly, this prevents `writeUtf8` from being inlined into callers, where `buffer` and `seq` could potentially monomorphise, and so instead manually specialises to recreate the opportunity to monomorphise those two, and allow their methods to inline?

The performance benefit looks kind of small, outside of the `nested… noUnsafe=false` cases, considering we now get 3 copies of the same code. It'd be more palatable if there were only two copies, along the safe/unsafe seam. Can you check and summaries the performance we'd get if we went with that?

franz1981 commented:
To summarize, the motivations behind this pr was indeed to:
1) makes writeUtf8 more resilient to be used on user code due to inlining budged exhaustion while reaching the PlatformDependent and CharSequence methods -> enlarged writeUtf8 enough to refill the inlining budget 
2) makes writeUtf8 more resilient to ByteBuf hierarchy, saving megamorphic calls -> specialized writeUtf8 in the 3 variants to cover all cases, decoupled by ByteBuf hierarchy

Results applying the solutions:
1) work well for nested calls but can cause megamorphic calls of CharSequence if the observed types > 2
2) works well and can be optimized with just 2 variant of the code too, merging the !PlatformDependent::hasUnsafe byte[] case into the AbstractByteBuf one (OpenJ9 users won't like it IMO)

My concern is if I should try to pursue 1) or not, given that most users just use String/StringBuilder...

chrisvest commented:
There's also `AsciiString` that makes its way into this method (in `StompSubframeEncoder.encodeFrame`), but that was the only case I could find in Netty itself (and it easily (I would think?) disappears if we call `toString` on those headers, or specialise for `AsciiString`).

franz1981 commented:
@chrisvest I've pushed some changes:

- handling `AsciiString` differently: similarly to writeUtf8, decoupling it from the `ByteBuf` hierarchy, but NOT enlarging it enough to not be inlined (the code just won't fit that purpose "naturally")
- adding some new benchmark method to show the difference with the version pre-PR

Soon the results

nitsanw commented:
@franz1981 one solution to mega-morphic callsites is to strip out the known main case, e.g.:
```
void doSomething(CharSequence cs) {
  // expected 60% of the time, inlining to known type locations should eliminate the `instanceof`
  if (cs instanceof String) {
   // we're monomorphic here
  }
  else {
    // we now extended the callsite type budget by one, so this may end up bi-morphic, or mega-morphic, but at least the main case is covered
  }
}
```

franz1981 commented:
this PR:
```
Benchmark                                                       (direct)  (noUnsafe)   Mode  Cnt      Score     Error  Units
Utf8EncodingBenchmark.byteBufUtilWriteUtf8AsciiString             false       false  thrpt   10  24442.071 ± 704.774  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8AsciiString             false        true  thrpt   10  25229.056 ± 339.069  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8AsciiString              true       false  thrpt   10  24563.689 ± 451.031  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8AsciiString              true        true  thrpt   10  14326.507 ± 105.995  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8CommonCharSequences     false       false  thrpt   10   3824.088 ± 28.556  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8CommonCharSequences     false        true  thrpt   10   2558.464 ± 26.097  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8CommonCharSequences      true       false  thrpt   10   3882.786 ± 71.585  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8CommonCharSequences      true        true  thrpt   10   1885.152 ± 19.069  ops/s
```
4.1:
```
Benchmark                                                      (direct)  (noUnsafe)   Mode  Cnt     Score     Error  Units
Utf8EncodingBenchmark.byteBufUtilWriteUtf8AsciiString             false       false  thrpt   10  13892.751 ±  97.401  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8AsciiString             false        true  thrpt   10  13389.432 ±  65.051  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8AsciiString              true       false  thrpt   10  15946.286 ± 244.559  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8AsciiString              true        true  thrpt   10   9161.213 ±  29.902  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8CommonCharSequences     false       false  thrpt   10   3539.553 ±  26.918  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8CommonCharSequences     false        true  thrpt   10   2833.111 ±   5.373  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8CommonCharSequences      true       false  thrpt   10   3667.355 ± 138.298  ops/s
Utf8EncodingBenchmark.byteBufUtilWriteUtf8CommonCharSequences      true        true  thrpt   10   1627.749 ± 942.380  ops/s
```
In summary: having a specialized `AsciiString` code path both reduce the chance to have > 2 `CharSequence` implementors and improve ASCII handling (for both write UTF-8 and write ASCII cases).

franz1981 commented:
@nitsanw 

re 
> // we're monomorphic here

I've just done it for `AsciiString`, but I'm not sure to do it for `String`/`StringBuilder` cases too or we would end having lot of copied methods...I mean, if it is for me, I will do it if the benefits are stable and better performance, but I understand if some of the Netty other committers won't like it too...

chrisvest commented:
@netty-bot test this please

Scottmitch commented:
@netty-bot test this please

chrisvest commented:
@franz1981 There are some checkstyle complaints. Please take a look!

njhill commented on buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:
```diff
@@ -723,32 +807,32 @@ public static ByteBuf writeAscii(ByteBufAllocator alloc, CharSequence seq) {
+                final int len = seq.length();
+                AbstractByteBuf byteBuf = (AbstractByteBuf) buf;
+                byteBuf.ensureWritable0(len);
+                int written;
```
nit: I know it was there before, but could get rid of `written` here and just use `len` directly in it's place

franz1981 commented:
@njhill @chrisvest @normanmaurer I've squashed the commits into one, addressed the checkstyle comments (and the Nick's one on `written`), but haven't yet fully used the good advice from @nitsanw .
Right now `writeUtf8` can see a maximum of 2 `CharSequence` implementors (`String`, `StringBuilder`), because `AsciiString` is being handled in a separate ad-hoc method: if we care about a specific case (eg `String` with `Unsafe`) I can add another separate (big) method to handle it, further increasing the chance to not getting monomorphic call-sites *if* we still care about the user code stability (reminder: I've enlarged the byte code size of each specialized method to replenish the inlining depth budget)

johnou commented:
@franz1981 looks like the average user running jdk8 and Netty would benefit from boosting -XX:MaxInlineLevel to something like 15?

franz1981 commented:
@johnou Absolutely yes, assuming that user code is performing many `ByteBuf` content manipulations (as @njhill has pointed clearly on past issues/PRs)

chrisvest commented:
@netty-bot test this please

nitsanw commented:
@franz1981 Looking at https://github.com/netty/netty/pull/10368#issuecomment-661811217 I see a regression for the CommonCharSequences on heap, no unsafe case (A is after, B is before):
```
                    | (direct) |(noUnsafe) | A    | B    |        
AsciiString         |    false |     false | 24.4 | 13.8 | kops/s 
AsciiString         |    false |      true | 25.2 | 13.3 | kops/s 
AsciiString         |     true |     false | 24.5 | 15.9 | kops/s 
AsciiString         |     true |      true | 14.3 |  9.1 | kops/s 
CommonCharSequences |    false |     false |  3.8 |  3.5 | kops/s 
CommonCharSequences |    false |      true |  2.5 |  2.8 | kops/s <- regression
CommonCharSequences |     true |     false |  3.8 |  3.6 | kops/s 
CommonCharSequences |     true |      true |  1.8 |  1.6 | kops/s
```
Worth looking into?

franz1981 commented:
@nitsanw 
> CommonCharSequences |    false |      true |  2.5 |  2.8 | kops/s

Good eyes! The no Unsafe case is getting a regression because of this request from @chrisvest on https://github.com/netty/netty/pull/10368#issuecomment-661019850

>  It'd be more palatable if there were only two copies, along the safe/unsafe seam

That I've "interpreted" by not optimizing the code path byte[] & no unsafe: I can restore that method I've removed to restore the perf benefit

njhill commented:
> looks like the average user running jdk8 and Netty would benefit from boosting -XX:MaxInlineLevel to something like 15?

@johnou I would even suggest higher e.g. 20 or more

franz1981 commented:
@nitsanw 
Restoring the array + no unsafe variant I'm getting:
```
Benchmark                                                      (direct)  (noUnsafe)   Mode  Cnt     Score    Error  Units
Utf8EncodingBenchmark.byteBufUtilWriteUtf8CommonCharSequences     false        true  thrpt   10  3330.168 ± 88.673  ops/s
```
that's a nice speedup over master IMO

nitsanw commented:
> > looks like the average user running jdk8 and Netty would benefit from boosting -XX:MaxInlineLevel to something like 15?
> 
> @johnou I would even suggest higher e.g. 20 or more

Do you have data supporting that? `MaxInlineDepth=15` has gone through some testing by OpenJDK team and is the new default for 14. With more inlining we can potentially get more code duplication -> larger code blobs -> more icache misses/code cache exhaustion, it's not a simple "more is better" knob to tune ;-)

franz1981 commented:
@nitsanw I would like to improve the `direct` and `no unsafe` method as well: I'll update my results here

franz1981 commented:
@nitsanw 

I've further improved the `no unsafe` and `direct` case too by using `internalNioBuffer`:
```
Utf8EncodingBenchmark.byteBufUtilWriteUtf8CommonCharSequences      true        true  thrpt   10  3510.155 ± 169.558  ops/s
```
while before it:
```
Utf8EncodingBenchmark.byteBufUtilWriteUtf8CommonCharSequences      true        true  thrpt   10   1885.152 ± 19.069  ops/s
```
I'm going to push this change too but probably need some help from @normanmaurer too while using `internalNioBuffer` because I'm not sure if is the right tool for the job

njhill commented:
> Do you have data supporting that?

@nitsanw no, it's just based on anecdotal observations, so thanks for the clarification/warning! I understand it is a trade-off but get the sense that the variance of optimal values across different apps is quite high, i.e. 15 may be optimal on average but there are certain cases where the benefit from larger values is huge. It would be neat if the JDK could be more adaptive somehow w.r.t. how deep it goes for a particular subtree... for example based on some some on-the-fly analysis of the benefit in terms secondary optimization potential (reorderings etc).

nitsanw commented:
@njhill the bounded size of the overall blob, callee methods size, and virtual callsites all play a role in limiting inlining, the hard limit is there as a "sane" boundary. The JVM, as of now, does not have a feedback loop from potential adverse effects (icache/iTLB misses etc) to revising inlining policy. I'm not sure any compiler out there does TBH. It's an interesting topic for a paper, worth checking if any were written. IIRC some work has been done on hiphop to reduce TLB misses through repacking of their code cache, which my be a relevant comparison.
Having said all that, if you have concrete examples where upping the limit is a win I would LOVE to hear more :-)

johnou commented:
@nitsanw I imagine it also boils down to the application (size and usage) not to mention hardware.. probably one of those knobs which needs to be tuned for the application / environment.

nitsanw commented:
Here's an interesting read: https://arxiv.org/pdf/1810.00905.pdf :-)

njhill commented:
@nitsanw isn't this PR itself a good example? If it wasn't for the artifical method inflation trick to reset the inlining budget then calling `ByteBufUtil#writeUtf8` from a deep app call stack would be very costly, to a degree I'd presume outweighs any adverse affects you mention. It's a step change in perf once it reaches the "leaf" methods containing the checks to reorder/hoist/coalese. You could see something similar in netty I think with a very long channel pipeline and a codec which performs many granular ByteBuf reads/writes.

I don't understand why bounding size of overall blob isn't sufficient, or at least should make it "safe" for the default depth limit to be larger, i.e. to avoid precluding long chains of trivially sized methods. Or maybe the size limit could depend on the current depth, getting smaller as the depth increases (but ultimately able to go deeper).

njhill commented:
Also just came across this which looks very relevant http://aleksandar-prokopec.com/resources/docs/prio-inliner-final.pdf in particular "adaptive decision thresholds" and "deep inlining trials".

franz1981 commented:
Thanks for all your helps :)

I've improved the test coverage for the ascii case (that has been improved as well) and now this one is in a completed state: please @normanmaurer, have fun :D

chrisvest commented:
@franz1981 You got some checkstyle violations in `ByteBufUtilTest`. (Build says there's one, but I think there's two)

franz1981 commented:
@chrisvest  Just tried a local run and it seems just 1, now fixed: which one is the second one 4 u?

chrisvest commented:
@franz1981 Hmm… I thought these two consecutive empty lines would also cause trouble, but apparently not? https://github.com/netty/netty/pull/10368/files#diff-21be031058403699535d72fa9ea7548fR45-R46

franz1981 commented:
@chrisvest Indeed is strange mmm let me remove that too and we're fine :)

franz1981 commented:
@chrisvest @normanmaurer @vietj Sadly I don't have some big machine around to try this, but would be nice to test this vs 4.1 on some Techempower text bench

chrisvest commented:
Checkstyle is at it again:

```
16:05:45 [INFO] Starting audit...
16:05:45 [ERROR] /code/microbench/src/main/java/io/netty/microbench/buffer/Utf8EncodingBenchmark.java:44:9: Redundant 'public' modifier. [RedundantModifier]
16:05:45 Audit done.
```

franz1981 commented:
@chrisvest I suppose I should better configure my IDE..strange...
![image](https://user-images.githubusercontent.com/13125299/91541773-227a8680-e91d-11ea-936d-59697bb6d47e.png)

franz1981 commented:
This one should be ready (despite the amount of duplicated code) @normanmaurer @chrisvest feel free to re-review this; the perf numbers are now pretty good and better/on-par with the original code

chrisvest commented:
@netty-bot test this please

chrisvest reviewed (approved):
Looks good. Thanks! 👍

chrisvest commented:
Hey @franz1981, can you give this a rebase and resolve the conflicts?

franz1981 commented:
@chrisvest done :+1:

chrisvest commented:
@netty-bot test this please

chrisvest commented:
Thanks a lot, @franz1981 !
